### PR TITLE
Update 7.0

### PR DESCRIPTION
**Pull request by Alluneve**
Hello again, and Welcome to Dawntrail!

with Dalamud in a state where plugin developers could start fixing their plugins, I've made 99% of MoActionPlugin work, 

- re-write the CanUseAction method using the Dalamud exposed ActionManager to simplify business logic, removing the need for some specific Sigs, only leaving ground target queue patch
- Fixed the **long-standing quest instance bug**
- introduced feature: Level sync Actions, skipping over stacks if your player's level is not high enough _a stack of target cure 2, target cure 1 will skip over cure 2 in level synced satasha_
- Fixed a bug with role actions and shared actions (like scholar/summoner res) bleeding between jobs an action stack on "Swiftcast" setup on white mage and Astrologian would conflict.
- Moved the UI mouse over logic to instead rely on STRUCTS, making the only self-maintained sig the ground target queueing patch.
- Applied the changes required for API 10
- Fixed the Ground target Queue patch signature, _with thanks to Cara from The Dalamud discord_ 

With lovely regards!

PS: total the pull request has grown larger than I realised, so at any point, if you need some extra information or have questions about anything in the pull request, please don't hesitate to ask any questions.

### CODE REVIEW

**Comment by Aqua-1994** [comment]
added a file with 12 scenarios I run when doing any change to the codebase, I'll look into adding UNIT tests later

File: MOAction/MOAction.cs [hunk 1]
```diff
@@ -239,109 +225,110 @@ private unsafe (Lumina.Excel.GeneratedSheets.Action action, GameObject target) G
...
+            bool selfOnlyTargetAction = !action.CanTargetFriendly && !action.CanTargetHostile && !action.CanTargetParty;
+            pluginLog.Verbose("Can {actionname} target: friendly - {friendly}, hostile  - {hostile}, party  - {party}, dead - {dead}, self - {self}", action.Name, action.CanTargetFriendly, action.CanTargetHostile, action.CanTargetParty, action.CanTargetDead, action.CanTargetSelf);
+            if (selfOnlyTargetAction)
+            {
+                pluginLog.Verbose("Can only use this action on player, setting player as target");
+                target = clientState.LocalPlayer;
+            }
+
+            //handoff to game native code, returns true if the actionManager declares that the action can be used on the target specified
+            bool gameCanUseActionResponse = ActionManager.CanUseActionOnTarget(action.RowId, (FFXIVClientStructs.FFXIV.Client.Game.Object.GameObject*)target.Address);
```
Comment (attickdoor) [inline MOAction/MOAction.cs:306]: I'm guessing this is the secret sauce that eliminates the need for the `unorthodox*` data? Those were always a conceptual pain.

**Comment by attickdoor** [comment]
Looks pretty good to me. Maybe a bit of tidying up blank lines and comments and such. Clearly I put off refactoring to more heavily use ClientStructs/etc for too long.

File: MOAction/MOAction.cs (hunk 1 above)
Comment (Alluneve) [reply MOAction/MOAction.cs:306]: that eliminated so so so so so much,
handing the decision over to the game in a way we want (native code), we use the game's decisionmaking to decide. killing all the unorthodox AND fixing the quest instance bug

**Comment by Alluneve** [comment]
> Looks pretty good to me. Maybe a bit of tidying up blank lines and comments and such. Clearly, I put off refactoring to more heavily use ClientStructs/etc for too long.

The comments are me explaining to myself what the fuck I was doing, so I remembered 😄

**Comment by attickdoor** [comment]
> you do have write access on the branch requesting merge, right?

Possibly, but my desktop still has yet to be setup. Clean up extra spacing and comments to yourself, and I'll LGTM™️ it.

**Comment by attickdoor** [comment]
> I'll look into adding UNIT tests later

I appreciate the thought, but when the target environment (a literal running executable) is so different from the dev environment (writing code and praying for the best), it's not obvious to me how to write useful tests. I guess one option would be a dev-compiled `/testmoaction` command that does various things in-game. That's a bit beyond me right now though, and might be better implemented as a Dalamud feature for indev plugins.

**Comment by Alluneve** [comment]
> > I'll look into adding UNIT tests later
> 
> I appreciate the thought, but when the target environment (a literal running executable) is so different from the dev environment (writing code and praying for the best), it's not obvious to me how to write useful tests. I guess one option would be a dev-compiled `/testmoaction` command that does various things in-game. That's a bit beyond me right now though, and might be better implemented as a Dalamud feature for indev plugins.

Dalamud Discord was talking about unit tests with everything being interfaced these days, so you can easily mock that stuff and get to call with fake data on the plugin's main hook and stuff. it was an ongoing conversation in plugin-dev

**Comment by kuujo** [comment]
FWIW I tried out the PR and it worked great for me!

**Comment by Alluneve** [comment]
@attickdoor I'll verify it all still works for the new hotfix

**Comment by Alluneve** [comment]
Update: still works with the hotfix, shipping-ready!

**Comment by Loucka88** [comment]
Plugin Updates are possible via Dalamud again, would love to see this live!

**Comment by Alluneve** [comment]
> Plugin Updates are possible via Dalamud again, would love to see this live!

this pull request is a request to merge changes from my fork into this repository, this has nothing to do with plugin updates in Dalamud.
As this plugin exists on the main branch, the plugin owner (attick) is responsible for this plugin not doing nefarious things (see the issues with g-shade dev injecting malware in their plugin). thus it also falls on them to make sure that the stuff I did in my pull request does not do bad shit or has code smells/bad code/...., because by merging this pull request they're taking personal responsibility for the stuff I wrote. As this is a big responsibility, please don't push or beg for speeding up of this pull request.

Attick has mentioned in discord: 

> i'm in the middle of moving and setting up my desktop is a relatively low priority. i doubt my dev setup will be online within the next few days.

thus please give them some space, **it'll come when it's ready.**

It's also why I'm not that keen on people bypassing proper plugin release flow and yoinking in-progress dev versions from my fork, as I do not want to be responsible for anyone's final fantasy account/character getting corrupted or any unforeseen bug.

**Comment by attickdoor** [comment]
> shipping ready!

I'm going to put it into staging or whatever and let it marinate for a couple of days. I greatly appreciate the update.

**Pull request merged**